flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/partitioner/FlinkFixedPartitioner.java

```diff
@@ -68,6 +78,13 @@ public int partition(T record, byte[] key, byte[] value, String targetTopic, int
 			partitions != null && partitions.length > 0,
 			"Partitions of the target topic is empty.");
 
-		return partitions[parallelInstanceId % partitions.length];
+		if (topicToFixedPartition.containsKey(targetTopic)) {
```

Comment:
This will only work if we don't restart the job, right? Because the state of the partitioner is lost when we have a failure.

Reply:
@aljoscha that is correct. I was hesitant in fixing that because it was a problem that had always been there even in the previous version of the FlinkFixedPartitioner, but AFAIK there hasn’t been complaints about it.

We could also fix that for good now. What do you think?

Reply:
How would we fix it for good?

Reply:
On second thought, passing in the runtime context wouldn’t work. To let user custom partitioners  be stateful, we’ll essentially need to make FlinkKafkaPartitioner a CheckpointedFunction, and let the FlinkKafkaProducerBase invoke the checkpoint methods.

We should be able to avoid breaking the user API by having empty base implementations for the checkpoint methods on the FlinkKafkaPartitioner.

Reply:
Maybe yes, but I don't think we can do it this quickly now. Adding stateful stuff has big implications and we should make sure to understand them all.

Reply:
It might be that the "fixed partitioner" is hard to implement or that its semantics are somewhat strange. Not sure though. What's the default Kafka behaviour? Maybe we should also have that as the default behaviour.

Reply:
@aljoscha yes, the semantics is a bit odd / needs some clarification before we move on. I've been having a go at implementing state checkpointing for the `FlinkFixedPartitioner` today, and for example one unclear case I bumped into was the following:
Subtask 1 writes to partition X for "some-topic"
Subtask 2 writes to partition Y for "some-topic"
On restore and say the sink is rescaled to DOP of 1, should the single subtask continue writing to partition X or Y for "some-topic"?

Regarding the default Kafka behaviour:
It's hash partitioning on the attached key for the records. I've also thought about using that as the default instead of the fixed partitioner; see the relevant discussion here: http://apache-flink-user-mailing-list-archive.2336050.n4.nabble.com/FlinkKafkaProducerXX-td16951.html